### PR TITLE
Allow searching for annotations in help

### PR DESCRIPTION
**Pull request by RedMser**
Show annotations in Search Help dialog, and add "Annotations Only" filter.

Also adds a new MemberAnnotation.svg icon.

image

### CODE REVIEW

File: editor/editor_help_search.h [hunk 1]
```diff
@@ -145,7 +147,7 @@ class EditorHelpSearch::Runner : public RefCounted {
 	void _match_item(TreeItem *p_item, const String &p_text);
 	TreeItem *_create_class_hierarchy(const ClassMatch &p_match);
 	TreeItem *_create_class_item(TreeItem *p_parent, const DocData::ClassDoc *p_doc, bool p_gray);
-	TreeItem *_create_method_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const String &p_text, const DocData::MethodDoc *p_doc);
+	TreeItem *_create_method_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const String &p_text, const DocData::MethodDoc *p_doc, const bool p_is_annotation);
```
Comment (akien-mga) [inline editor/editor_help_search.h:150]: We don't use `const` for scalar parameter types.

File: editor/editor_help_search.cpp [hunk 2]
```diff
@@ -588,7 +599,13 @@ TreeItem *EditorHelpSearch::Runner::_create_method_item(TreeItem *p_parent, cons
 		}
 	}
 	tooltip += ")";
-	return _create_member_item(p_parent, p_class_doc->name, "MemberMethod", p_doc->name, p_text, TTRC("Method"), "method", tooltip);
+
+	if (p_is_annotation) {
+		// TODO: Different icon?
```
Comment (akien-mga) [inline editor/editor_help_search.cpp:604]: Yeah it should likely have an icon derived from `@`.

Comment (RedMser) [reply editor/editor_help_search.cpp]: Having just the `@` looks a bit weird when doubled with the annotation's name... 

image

I'm not a graphics designer, so I wouldn't mind if someone else comes up with a better icon :)

Comment (YuriSizov) [reply editor/editor_help_search.cpp]: It's fine for now

File: editor/editor_help_search.cpp [hunk 3]
```diff
@@ -575,7 +586,7 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
 	return item;
 }
 
-TreeItem *EditorHelpSearch::Runner::_create_method_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const String &p_text, const DocData::MethodDoc *p_doc) {
+TreeItem *EditorHelpSearch::Runner::_create_method_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const String &p_text, const DocData::MethodDoc *p_doc, bool p_is_annotation) {
```
Comment (YuriSizov) [inline editor/editor_help_search.cpp:589]: I would create a dedicated method instead. Signal also uses MethodDoc and is formatted similarly, yet it has a dedicated method. A little code duplication doesn't hurt here (and it's better to instead factor out similar bits of unique methods to a dedicated helper, if this gets too obnoxious).

**Comment by YuriSizov** [comment]
Seems to work fine, but needs the aforementioned refactoring. It's a bit annoying that inputting "@" doesn't show all annotation, but that's only because we filter members starting with a 2-character-long input. Oh well.

As for the icon visually colliding with the name issue, we can split the name argument of `_create_member_item` into the actual name and the display name, and strip the "at" sign for the display name. (Matching should work regardless of user inputting "@" or not, so the match name should be the actual name).

File: editor/editor_help_search.cpp [hunk 4]
```diff
@@ -576,37 +604,20 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
...
-		}
-	}
-	tooltip += ")";
+	String tooltip = _build_method_tooltip(p_class_doc, p_doc);
 	return _create_member_item(p_parent, p_class_doc->name, "MemberSignal", p_doc->name, p_doc->name, TTRC("Signal"), "signal", tooltip);
 }
 
+TreeItem *EditorHelpSearch::Runner::_create_annotation_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const String &p_text, const DocData::MethodDoc *p_doc) {
+	String tooltip = _build_method_tooltip(p_class_doc, p_doc);
+	return _create_member_item(p_parent, p_class_doc->name, "Annotation", p_doc->name, p_text, TTRC("Annotation"), "annotation", tooltip);
```
Comment (YuriSizov) [inline editor/editor_help_search.cpp:618]: Just realized that the icon should be `MemberAnnotation` for consistency.

**Comment by RedMser** [comment]
> [...] we filter members starting with a 2-character-long input

Yeah it seems like there's a fairly big lag spike (around 1-2 seconds) when building a tree for many results.

But since there are only few annotations compared to other member types, I feel it's alright to add a special case for if you type only an `@` in the search logic.

> [...] strip the "at" sign for the display name

The code already works like that, so I only had to remove the `@` from `p_text` to change display text, and searching still works like before.

**Review by YuriSizov — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**